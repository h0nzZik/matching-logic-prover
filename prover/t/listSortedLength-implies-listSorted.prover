claim \implies( \and( listSortedLength( variable("H") { ArrayIntInt }
                                      , variable("X") { Int }
                                      , variable("K") { Set }
                                      , variable("MIN") { Int }
                                      , variable("Length") { Int }
                                      , .Patterns
                                      )
                    , .Patterns
                    )
              , \and( listSorted( variable("H") { ArrayIntInt }
                                , variable("X") { Int }
                                , variable("K") { Set }
                                , variable("MIN") { Int }
                                , .Patterns
                                )
                    , .Patterns
                    )
              )

strategy search-bound(3)

/*
strategy kt ; ( (right-unfold-Nth(0, 0) ; simplify ; direct-proof )
              | (simplify ; direct-proof)
              | (right-unfold-Nth(0, 1) ; simplify ; direct-proof )
              )
*/
