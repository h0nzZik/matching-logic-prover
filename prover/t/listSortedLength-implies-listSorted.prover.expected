<prover>
  <k>
    .
  </k>
  <goals>
    <goal>
      <id>
        root
      </id>
      <active>
        true
      </active>
      <parent>
        .
      </parent>
      <claim>
        \implies ( \and ( listSortedLength ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "K" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "Length" ) { Int } , .Patterns ) , .Patterns ) , \exists { .Patterns } \and ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "K" ) { SetInt } , variable ( "MIN" ) { Int } , .Patterns ) , .Patterns ) )
      </claim>
      <strategy>
        success
      </strategy>
      <trace>
        .
      </trace>
    </goal>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , .Patterns } \iff-lfp ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , variable ( "X_I" ) { Int } , .Patterns } \and ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I" ) { Int } , variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X_I" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , \equals ( variable ( "VAL_I" ) { Int } , select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( variable ( "VAL_I" ) { Int } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , variable ( "LENGTH" ) { Int } , .Patterns } \iff-lfp ( listSortedLength ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , variable ( "LENGTH" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , \equals ( variable ( "LENGTH" ) { Int } , 0 ) , .Patterns ) , \exists { variable ( "F_I" ) { SetInt } , variable ( "LENGTH_I" ) { Int } , variable ( "VAL_I" ) { Int } , variable ( "X_I" ) { Int } , .Patterns } \and ( listSortedLength ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I" ) { Int } , variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , variable ( "LENGTH_I" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X_I" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , \equals ( variable ( "VAL_I" ) { Int } , select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( variable ( "VAL_I" ) { Int } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , gt ( variable ( "LENGTH" ) { Int } , 0 , .Patterns ) , \equals ( variable ( "LENGTH_I" ) { Int } , minus ( variable ( "LENGTH" ) { Int } , 1 , .Patterns ) ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      symbol listSorted ( ArrayIntInt , Int , SetInt , Int , .Sorts ) : Bool
    </declaration> <declaration>
      symbol listSortedLength ( ArrayIntInt , Int , SetInt , Int , Int , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
